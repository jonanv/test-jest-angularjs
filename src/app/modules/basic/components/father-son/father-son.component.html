<p>father-son works!</p>
