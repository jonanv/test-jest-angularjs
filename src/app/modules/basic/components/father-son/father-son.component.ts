import { Component, EventEmitter, Input, Output } from '@angular/core';

// Interfaces
import { Client } from '../../interfaces/client.interface';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styles: ``
})
export class FatherSonComponent {

  @Input() public client!: Client | undefined;
  @Output() public onDeleteClientEmitter: EventEmitter<undefined> = new EventEmitter<undefined>();
  @Output() public onUpdateClientEmitter: EventEmitter<Client> = new EventEmitter<Client>();

  public onUpdateClient(id: number): void {
    if (!this.client) return;
    this.client = { ...this.client, id };
    this.onUpdateClientEmitter.emit(this.client);
  }

  public onDeleteClient(): void {
    this.client = undefined;
    this.onDeleteClientEmitter.emit();
  }

}
