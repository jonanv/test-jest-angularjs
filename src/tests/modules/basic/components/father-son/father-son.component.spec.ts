import { ComponentFixture, TestBed } from '@angular/core/testing';

// Components
import { FatherSonComponent } from '../../../../../app/modules/basic/components/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [FatherSonComponent]
    })
    .compileComponents();

    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement as HTMLElement;
    jest.clearAllMocks();
  });

  test('Debe de crearse el componente', () => {
    expect(component).toBeTruthy();
  });

  test('Debe de hacer match con el snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('No deben aparecer botones si no hay cliente', () => {
    let buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0);
  });

  test('Deben de aparecer 2 botones si hay cliente', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    let buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2);
  });

  test('Deben de match con el snapshot si hay cliente', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    expect(compiled).toMatchSnapshot();
  });

  test('Debe de emitir onDeleteClientEmitter con el boton de "Borrar cliente"', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onDeleteClientEmitter, 'emit');
    let btnOnDeleteClient = compiled.querySelector('[data-test="btn-delete-client"]');
    // console.log(btnOnDeleteClient?.innerHTML);
    btnOnDeleteClient?.dispatchEvent(new Event('click'));

    expect(component.onDeleteClientEmitter.emit).toHaveBeenCalled();
  });

  test('Debe de emitir onUpdateClientEmiiter con el boton de "Cambiar id"', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onUpdateClientEmitter, 'emit');
    let btnOnUpdateClient = compiled.querySelector('[data-test="btn-update-client"]');
    // console.log(btnOnUpdateClient?.innerHTML);
    btnOnUpdateClient?.dispatchEvent(new Event('click'));

    expect(component.onUpdateClientEmitter.emit).toHaveBeenCalledWith({
      id: 5,
      name: 'Juan'
    });
  });

  test('Debe de emitir el onUpdateCliente con el ID especificado si hay un cliente', () => {
    jest.spyOn(component.onUpdateClientEmitter, 'emit');

    // Cuando el cliente NO existe
    component.onUpdateClient(10);
    expect(component.onUpdateClientEmitter.emit).not.toHaveBeenCalled();

    // Cuando el cliente existe
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    component.onUpdateClient(10);

    expect(component.onUpdateClientEmitter.emit).toHaveBeenCalledWith({
      id: 10,
      name: 'Juan'
    });
  });

});
