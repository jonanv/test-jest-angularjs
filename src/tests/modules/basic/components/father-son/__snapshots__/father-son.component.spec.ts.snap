// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent Debe de hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  <div
    class="d-flex gap-2"
  >
    
    
  </div>
</div>
`;

exports[`FatherSonComponent Deben de match con el snapshot si hay cliente 1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
    {
  "id": 1,
  "name": "Juan"
}
  </code>
  <br />
  <br />
  <div
    class="d-flex gap-2"
  >
    <button
      class="btn btn-primary"
      data-test="btn-update-client"
    >
       Cambiar ID 
    </button>
    
    <button
      class="btn btn-danger"
      data-test="btn-delete-client"
    >
       Borrar cliente 
    </button>
    
  </div>
</div>
`;
